### PR TITLE
[SYCL] [FPGA] Update handler of variadic template argument list in fpga_utils.hpp to support more types other than int32_t 

### PR DESCRIPTION
**Pull request by shuoniu-intel**
This patch is a prerequisite for the upcoming FPGA latency control feature.

### CODE REVIEW

**Review by dm-vodopyanov — DISMISSED** [review]
LGTM

**Comment by bader** [comment]
> to support more types other than int32_t

Is there any test checking for that?

**Comment by shuoniu-intel** [comment]
> > to support more types other than int32_t
> 
> Is there any test checking for that?

Not yet, the upcoming latency control header change will use other types. But as for this patch, there's no functional difference in fpga.lsu.hpp.

**Comment by bader** [comment]
> Do you think I should add test in this patch?

Yes. Otherwise it's quite likely that this functionality might be unintentionally broken by future patches (e.g. code refactoring).

**Comment by shuoniu-intel** [comment]
> > Do you think I should add test in this patch?
> 
> Yes. Otherwise it's quite likely that this functionality might be unintentionally broken by future patches (e.g. code refactoring).

Sure, could you point me where to add the test?

**Comment by bader** [comment]
Here are common guidelines: https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#tests-development
@intel/llvm-reviewers-runtime should be able to provide more detailed instructions how to write a test for this feature.

**Comment by romanovvlad** [comment]
> 
> 
> Here are common guidelines: https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#tests-development
> @intel/llvm-reviewers-runtime should be able to provide more detailed instructions how to write a test for this feature.

It seems we need a test which verifies produced IR here. @sergey-semenov  Could you please confirm?

**Comment by sergey-semenov** [comment]
> > Here are common guidelines: https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#tests-development
> > @intel/llvm-reviewers-runtime should be able to provide more detailed instructions how to write a test for this feature.
> 
> It seems we need a test which verifies produced IR here. @sergey-semenov Could you please confirm?

I think using `static_assert` to verify the correctness of the variadic templates with the newly supported types would be preferable here (would save us a run of FileCheck).

**Comment by shuoniu-intel** [comment]
> > > Here are common guidelines: https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#tests-development
> > > @intel/llvm-reviewers-runtime should be able to provide more detailed instructions how to write a test for this feature.
> > 
> > 
> > It seems we need a test which verifies produced IR here. @sergey-semenov Could you please confirm?
> 
> I think using `static_assert` to verify the correctness of the variadic templates with the newly supported types would be preferable here (would save us a run of FileCheck).

If the only test we need here is a cpp file that includes this header, where should I put the test file? I don't see the guidelines link above covers it.

**Comment by bader** [comment]
`sycl/test` directory is the right place.
`sycl/test/basic_tests/id_ctad.cpp` - good example of using `static_assert`.

**Comment by shuoniu-intel** [comment]
I have created a new test - sycl/test/fpga_tests/arbitrary_template_arg.cpp. It's not a static_assert check but a host run check.
However I'm not sure why Jenkins/Precommit test times out on Windows, I don't think it's related to my new commit. 
(http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FPrecommit_Check_User/detail/Precommit_Check_User/7484/pipeline)

**Comment by bader** [comment]
> I have created a new test - sycl/test/fpga_tests/arbitrary_template_arg.cpp. It's not a static_assert check but a host run check.
> However I'm not sure why Jenkins/Precommit test times out on Windows, I don't think it's related to my new commit.
> (http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FPrecommit_Check_User/detail/Precommit_Check_User/7484/pipeline)

I would prefer to have the test similar to `sycl/test/basic_tests/id_ctad.cpp` with static_assert checks.
According to my understanding, to validate the changes made by this patch there is no need to run the compiled test, so we can reduce the test time and make it more reliable.
Does it make sense?

**Comment by shuoniu-intel** [comment]
> I would prefer to have the test similar to `sycl/test/basic_tests/id_ctad.cpp` with static_assert checks.
> According to my understanding, to validate the changes made by this patch there is no need to run the compiled test, so we can reduce the test time and make it more reliable.
> Does it make sense?

It makes sense to remove the compiled test :).

As for the test I added, I want to check both type support (int and enum class) and arbitrary order support of template argument list (users don't have to specify all template arguments, and template arguments can be specified in arbitrary order). A run check covers both ends.

**Comment by sergey-semenov** [comment]
> > I would prefer to have the test similar to `sycl/test/basic_tests/id_ctad.cpp` with static_assert checks.
> > According to my understanding, to validate the changes made by this patch there is no need to run the compiled test, so we can reduce the test time and make it more reliable.
> > Does it make sense?
> 
> It makes sense to remove the compiled test :).
> 
> As for the test I added, I want to check both type support (int and enum class) and arbitrary order support of template argument list (users don't have to specify all template arguments, and template arguments can be specified in arbitrary order). A run check covers both ends.

I'm not quire sure how the runtime check here does anything a compile-time check wouldn't be able to do. Something like this should check the same behaviour as what you have now.

```
template <int ExpectedIntValue, enum_class ExpectedEnumValue, class... _Params> void func {
  static_assert(sycl::INTEL::_GetValue<test_int<0>, _Params...>::value == ExpectedIntValue);
  static_assert(sycl::INTEL::_GetValue<test_enum_class<enum_class::first>, _Params...>::value == ExpectedEnumValue);
}

int main() {
  func<0, enum_class::first>();
  func<1, enum_class::first, test_int<1>>();
  ...
}
```

**Comment by shuoniu-intel** [comment]
Changed my test to use compile-time static_assert.

File: sycl/test/fpga_tests/arbitrary_template_arg.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      sycl::INTEL::_GetValue<testEnum<enumClass::first>, _Params...>::value ==
+      ExpectedEnumValue);
+}
+
+int main() {
+  func<0, enumClass::first>();
+  func<1, enumClass::first, testInt<1>>();
+  func<0, enumClass::second, testEnum<enumClass::second>>();
+  func<1, enumClass::second, testInt<1>, testEnum<enumClass::second>>();
+  func<1, enumClass::second, testEnum<enumClass::second>, testInt<1>>();
```
Comment (dm-vodopyanov) [inline sycl/test/fpga_tests/arbitrary_template_arg.cpp:37]: `burst_coalesce_impl`, `cache`, `prefetch_impl` and `statically_coalesce_impl` should be covered by tests too.

Comment (shuoniu-intel) [reply sycl/test/fpga_tests/arbitrary_template_arg.cpp:37]: This new test is created only for fpga_utils.hpp.
I believe functionality of fpga_lsu.hpp is tested by `[2021-07-07T18:51:41.321Z] PASS: SYCL :: Basic/fpga_tests/fpga_lsu.cpp (59 of 530)` in the compiled test (Jenkins/Precommit: http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FLin%2FLLVM_Test_Suite/detail/LLVM_Test_Suite/5893/pipeline).

Comment (dm-vodopyanov) [reply sycl/test/fpga_tests/arbitrary_template_arg.cpp:37]: OK, sounds good to me.

**Review by dm-vodopyanov — APPROVED** [review]

File: sycl/test/fpga_tests/arbitrary_template_arg.cpp (hunk 1 above)
Comment (shuoniu-intel) [reply sycl/test/fpga_tests/arbitrary_template_arg.cpp:37]: Also, I just found llvm/sycl/test/extensions/fpga.cpp tests all FPGA interfaces including fpgs_lsu. See line 13-25 where `burst_coalesce_impl`, `cache`, `prefetch_impl`, and `statically_coalesce_impl` are tested.

**Pull request merged**